### PR TITLE
IDENTITY-2821 authentication.endpoint

### PR DESCRIPTION
**Pull request by thariyarox**

### CODE REVIEW

File: components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+					reader.close();
+				}
+				if (inputStream != null) {
+					inputStream.close();
+				}
+				if (outputStream != null) {
+					outputStream.close();
+				}
+			} catch (IOException e) {
+				log.error("Closing stream for " + url + " failed");
```
Comment (muthulee) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java]: The exception is not not logged. It is swollen.

Comment (thariyarox) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java]: Fixed in PR https://github.com/wso2-dev/carbon-identity/pull/56/files

File: components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantDataManager.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        return tenantDomainList;
+    }
+
+    /**
+     * Set the updated tenant domains list
+     * @param dataList List of active tenant domains
+     */
+    public static void setTenantDataList(String dataList) {
+
+        if (!initialized) {
```
Comment (prabathabey) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantDataManager.java]: Without initializing "TenantManager" at various levels, what if we write the logic in a way that it makes sure "TenantManager" is always inititalized before we invoke any of its methods? If there's any instance where there's a possibility that any of its methods would be invoked before the "TenantManager" gets successfully inititalized, you can perhaps add some assertions to validate the state.

File: components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+			reader = new BufferedReader(new InputStreamReader(inputStream));
+			StringBuilder builder = new StringBuilder();
+			String line;
+
+			while (StringUtils.isNotEmpty(line = reader.readLine())) {
+				builder.append(line);
+			}
+			response = builder.toString();
+		} catch (IOException e) {
+			log.error("Sending " + HTTP_POST + " request to URL : " + url + "failed.", e);
```
Comment (prabathabey) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java]: Any reason why we do not throw the exception upwards to the upstream logic?

Comment (johannnallathamby) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantMgtAdminServiceClient.java]: The idea is that we are gracefully recovering from the server error at this point by continuing with empty String, which means the tenant dropdown will not be populated with tenants, and user has to enter tenant domain manually when doing SSO with Identity Server, which is actually the default case.

File: components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantDataManager.java (hunk 2 above)
Comment (thariyarox) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantDataManager.java]: There are 3 entry points to the TenantDataManager based on different use cases. 

1. When a user visits SSO Login page, it needs to decide whether to display the tenant list dropdown. For that it calls following method. 
public static boolean isTenantListEnabled()

2. From the server side, a new tenant is added or existing tenant activated/deactivated. Then it calls following method to set the modified tenant domains list. 
public static void setTenantDataList(String dataList)

3. When the tenant domains list is empty in the authenticationendpoint webapp and a client visits SSO login page when the dropdown feature is enabled, it calls the TenantMgtAdmin service and retrieve the tenant domains list by calling following method.
public static List<String> getAllActiveTenantDomains() 

Based on above use cases, the entry point to the TenantDataManager differs. So to facilitate all of those entry points, initialization is checked in each scenario. There is no order for those 3 scenarios getting executed and also at a time not all 3 methods are getting called.

Comment (thariyarox) [inline components/identity/org.wso2.carbon.identity.application.authentication.endpoint/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/TenantDataManager.java]: Hi Prabath,

Thank you very much for your suggestion on improving the code. I added a ServletContextListener class to the webapp and implemented the  contextInitialized() method initializing the TenantDataManager. That way I could eliminate initialization of the 3 different places I mentioned in previous comment. Created the pull request in [1] for this fix. Thanks a lot again for your valuable input. 

[1] https://github.com/wso2/carbon-identity/pull/39

**Pull request merged**